### ISSUE TITLE
compiling problem Unbound type constructor Yojson.Basic.t 

### ISSUE THREAD

**Issue by glucksfall** [Kappa-Dev/KaSim#552]
Hello,

I'm having an error while trying to compile KaSim from the github last commit
```
File "dataStructures/jsonUtil.mli", line 28, characters 25-39:
Error: Unbound type constructor Yojson.Basic.t
```

I try with ocaml 4.05.0 and 4.02.3

Thank you

**Comment by hmedina** [comment]
What version of Yojson do you have installed?
I'm using Opam to manage the KaSim installation, and this commit 68d0f0be74b97f70087d15779152dbf51d6f147f compiled fine for me with Yojson 1.6 and 1.7 ([1.6 introduced the type `t`](https://github.com/ocaml-community/yojson/blob/master/CHANGES.md)).

**Comment by feret** [comment]
I got the same problem last week and I could not have solved the problem without Pirbo's help. 

Please upgrade yojson to version 1.6.0. 
->     opam upgrade yojson.1.6.0

Then either it works or not. 
-> On my MacOS machines, it was failing to compile some lib. 
I basically reset my opam repository (by removing .opam) and restarted the installation. 
-> On my ubuntu machine, Bubble wrap (a tool to make sandboxing) was missing. 
I had to install it by hand (it was not so easy, but I managed), before upgrading yojson. 

There may be more direct ways to may Yojson 1.6 work, but at least it worked. 

Cheers.

**Comment by hmedina** [comment]
On my frankensystem (WSL), I also had to nuke .opam as the change to bubblewrap usage broke a lot. I then [reinitialized Opam with bubblewrap disabled](http://opam.ocaml.org/doc/FAQ.html#Why-does-opam-require-bwrap).

**Comment by glucksfall** [comment]
Indeed the problem comes from an outdated yojson library. I tried to install 1.6.0 but opam refuse to install it because opam doesn't find it, and I'm stuck with yojson 1.4.1 and kasim4 RC1 compiles without a problem. I'll try @feret steps, but for now, I have more urgent priorities.

Thank for everything. Probably would be good to close the issue because isn't about KaSim itself, if you're ok with that.

**Comment by glucksfall** [comment]
I have bubblewrap (or at least, the man bwrap command return the manual of `bwrap - container setup utility`). I deleted .opam folder and did opam init again and here is the interesting part:

```
[NOTE] The repository 'default' will be *permanently* redirected to
       https://opam.ocaml.org/1.2.2 (opam-version < "2.0~")
```

Apparently, I can't have opam 2.0, which it's the version that needs bubblewrap. Also, this message `Use wisely, broken Makefiles that run rm -rf / do happen.` is really worrying (from http://opam.ocaml.org/doc/FAQ.html#Why-does-opam-require-bwrap). I think I'll wait until opam get it shit together.

**Comment by pirbo** [comment]
Sorry for the mess, OCaml community clearly moved too fast by releasing opam2 and deprecating opam1 package repository the same day!
Opam2 makes its way in ubuntu as well as in other linux distribution, it will be in the next release but indeed it was premature to say, we stop supporting opam1.

There is clearly an issue, You all come to me 1 by 1 after failing to update!!! 
KaSim is the victim not the author problem :-). I know It has to be mitigated in KaSim anyway. I can't made my mind on what is the least bad way...

**Comment by feret** [comment]
I am pretty sure that magic @pirbo will come with a solution soon.

**Comment by glucksfall** [comment]
Ok... I have a solution. 

I uninstalled ocaml and opam and reinstalled it from https://github.com/ocaml/ocaml and https://github.com/ocaml/opam. I checkout opam to 2.0 and deleted from .profile the OPAM configuration one-liner (if I could describe as it). I rebooted my laptop and followed compilation instructions from both repositories. Then I installed all dependencies with opam (dune, yojson, lwt, logs, num, re, fmt), then make to compile KaSim 4.1

The problem still is ocaml and opam are outdated within the ubuntu (18.04) repositories. Anyway, I'll close this issue.

Best regards to all
